api/src/wfl/module/covid.clj

```diff
@@ -88,42 +70,96 @@
+   return a [type item] pair to be written into the parent table."
+  (fn [tx id executor-request] (:name executor-request)))
 
 (defmulti update-executor!
   "Update the executor with the `source`"
   (fn [source executor] (:type executor)))
 
 (defmulti load-executor!
-  "Load the workload `executor`."
-  (fn [workload] (:executor-type workload)))
+  "Use `tx` to load the workload executor with `executor_type`."
+  (fn [tx workload] (:executor_type workload)))
 
 ;; sink operations
 (defmulti create-sink!
-  "Create and write the sink to persisted storage"
-  (fn [sink-request] (:name sink-request)))
+  "Use `tx` and workload `id` to write the sink to persisted storage and
+   return a [type item] pair to be written into the parent table."
+  (fn [tx id sink-request] (:name sink-request)))
 
 (defmulti update-sink!
   "Update the sink with the `executor`"
   (fn [executor sink] (:type sink)))
 
 (defmulti load-sink!
-  "Load the workload `sink`."
-  (fn [workload] (:sink-type workload)))
+  "Use `tx` to load the workload sink with `sink_type`."
+  (fn [tx workload] (:sink_type workload)))
+
+(defn ^:private add-continuous-workload-record
+  "Use `tx` and workload `id` to create a \"ContinuousWorkload\" instance and
+  return the ID of the ContinuousWorkload."
+  [tx id {:keys [source sink executor] :as _request}]
+  (let [set-details "UPDATE
+                         ContinuousWorkload
+                     SET
+                         source_type   = ?::source,
+                         executor_type = ?::executor,
+                         sink_type     = ?::sink
+                     WHERE
+                         id = ? "
+        src-exc-snk [(create-source! tx id source)
+                     (create-executor! tx id executor)
+                     (create-sink! tx id sink)]
+        items       (->> (map second src-exc-snk)
+                         (zipmap [:source_items :executor_items :sink_items])
+                         (jdbc/insert! tx :ContinuousWorkload)
+                         first
+                         :id)]
+    (jdbc/execute! tx (concat [set-details] (map first src-exc-snk) [items]))
+    items))
+
+(defn ^:private add-workload-record [tx request]
+  "Use `tx` to create a workload `record` for `request` and return the id of the
+   new workload."
+  (letfn [(combine-labels [labels]
+            (->> (mapv request [:pipeline :project])
+                 (map str ["pipeline:" "project:"])
+                 (concat labels)
+                 set
+                 sort
+                 vec))]
+    (-> (update request :labels combine-labels)
+        (select-keys [:creator :watchers :labels :project])
+        (merge (select-keys (wfl/get-the-version) [:commit :version]))
+        (assoc :executor "" :output "" :release "" :wdl "" :uuid (UUID/randomUUID))
+        (->> (jdbc/insert! tx :workload) first :id))))
+
+;; TODO: implement COVID workload creation
```

Comment:
Still a TODO?

Reply:
Yeah maybe not lol. There's still connecting up the request validation but that's not exactly covered by this to do.